File: internal/db/memory.go
package db

import (
	"context"
	"database/sql"
	"fmt"
	"regexp"
	"strings"
)

// Memory source kinds. A memory note originates from the legacy cross-agent
// SSOT (~/.dotfiles/memory/user), the explicit assist-mem ledger, CC-native
// auto-memory directories (~/.claude/projects/<project>/memory), or generated
// canonical rows. The source column lets the single memory table hold multiple
// data sources and the UI filter between them.
const (
	// SourceCrossAgent is the existing cross-agent user-memory SSOT.
	SourceCrossAgent = "cross-agent"
	// SourceAssistMem is the explicit assist-mem JSONL ledger.
	SourceAssistMem = "assist-mem"
	// SourceCCNative is CC-native auto-memory scanned across project dirs.
	SourceCCNative = "cc-native"
	// SourceCanonical is generated current-memory output with raw provenance.
	SourceCanonical = "canonical"
)

// Memory is one user-memory note: a markdown file under a memory data source
// (the cross-agent SSOT ~/.dotfiles/memory/user/*.md, or CC-native
// auto-memory ~/.claude/projects/<project>/memory/*.md) with optional YAML
// frontmatter and a body. Like Skill it is slowly-changing reference data
// that lives in its own dimension table and is populated by a MemorySyncer,
// never by the session sync path. The store is read-only for callers other
// than the syncer; PG/DuckDB receive rows via the SQLite mirror.
type Memory struct {
	RelPath       string `json:"rel_path"`
	Source        string `json:"source"`
	Title         string `json:"title"`
	Date          string `json:"date"`
	ProblemType   string `json:"problem_type"`
	Type          string `json:"type"`
	Status        string `json:"status"`
	OriginSession string `json:"origin_session"`
	// OriginProject is the project a note belongs to ("" = the General bucket:
	// user-global or cross-project notes). It drives the /memories project facet.
	OriginProject        string    `json:"origin_project"`
	FeedbackVote         string    `json:"feedback_vote"`
	FeedbackComment      string    `json:"feedback_comment"`
	FeedbackStatus       string    `json:"feedback_status"`
	CanonicalCoveredRefs string    `json:"canonical_covered_refs"`
	CanonicalProvenance  string    `json:"canonical_provenance"`
	Body                 string    `json:"body,omitempty"`
	BodyTokens           int       `json:"body_tokens"`
	SourceMtime          int64     `json:"source_mtime"`
	SyncedAt             string    `json:"synced_at"`
	LLMEmbedding         []float32 `json:"-"`
	LLMEmbeddingDim      int       `json:"-"`
}

// MemoryFilter narrows a memory listing. Empty fields = no filter. Q is a
// full-text query over the note body (FTS5 MATCH on SQLite, dialect-
// specific elsewhere). Source filters by data source (cross-agent vs
// cross-agent, assist-mem, cc-native, or canonical).
type MemoryFilter struct {
	Source         string
	ProblemType    string
	Type           string
	Status         string
	OriginSession  string
	OriginProject  string
	FeedbackVote   string
	FeedbackStatus string
	Q              string
}

// memoryCols lists the memory columns in the canonical order shared by
// every backend's scan helper.
const memoryCols = `rel_path, source, title, date, problem_type, type, status,
	origin_session, origin_project, feedback_vote, feedback_comment,
	feedback_status, canonical_covered_refs, canonical_provenance, body,
	body_tokens, source_mtime, synced_at`

const memoryEmbeddingCols = `llm_embedding, llm_embedding_dim`

func scanMemory(rows *sql.Rows) (Memory, error) {
	var m Memory
	if err := rows.Scan(
		&m.RelPath, &m.Source, &m.Title, &m.Date, &m.ProblemType, &m.Type,
		&m.Status, &m.OriginSession, &m.OriginProject, &m.FeedbackVote,
		&m.FeedbackComment, &m.FeedbackStatus, &m.CanonicalCoveredRefs,
		&m.CanonicalProvenance, &m.Body, &m.BodyTokens, &m.SourceMtime,
		&m.SyncedAt,
	); err != nil {
		return Memory{}, err
	}
	return m, nil
}

func scanMemoryWithEmbedding(rows interface {
	Next() bool
	Scan(...any) error
	Err() error
}) ([]Memory, error) {
	var out []Memory
	for rows.Next() {
		var m Memory
		var data []byte
		var dim int
		if err := rows.Scan(
			&m.RelPath, &m.Source, &m.Title, &m.Date, &m.ProblemType, &m.Type,
			&m.Status, &m.OriginSession, &m.OriginProject, &m.FeedbackVote,
			&m.FeedbackComment, &m.FeedbackStatus, &m.CanonicalCoveredRefs,
			&m.CanonicalProvenance, &m.Body, &m.BodyTokens, &m.SourceMtime,
			&m.SyncedAt, &data, &dim,
		); err != nil {
			return nil, err
		}
		if len(data) > 0 || dim > 0 {
			vector, err := DecodeEmbedding(data, dim)
			if err != nil {
				return nil, err
			}
			m.LLMEmbedding = vector
			m.LLMEmbeddingDim = dim
		}
		out = append(out, m)
	}
	return out, rows.Err()
}

// ListMemories returns memory notes, optionally filtered by frontmatter
// fields or a full-text query over the body, ordered by date descending
// then rel_path for stable display. The body is included on each row so
// the listing view can render snippets without a second round trip.
func (db *DB) ListMemories(
	ctx context.Context, f MemoryFilter,
) ([]Memory, error) {
	var preds []string
	var args []any
	if f.Source != "" {
		preds = append(preds, "m.source = ?")
		args = append(args, f.Source)
	}
	if f.ProblemType != "" {
		preds = append(preds, "m.problem_type = ?")
		args = append(args, f.ProblemType)
	}
	if f.Type != "" {
		preds = append(preds, "m.type = ?")
		args = append(args, f.Type)
	}
	if f.Status != "" {
		preds = append(preds, "m.status = ?")
		args = append(args, f.Status)
	}
	if f.OriginSession != "" {
		preds = append(preds, "m.origin_session = ?")
		args = append(args, f.OriginSession)
	}
	if f.OriginProject != "" {
		preds = append(preds, "m.origin_project = ?")
		args = append(args, f.OriginProject)
	}
	if f.FeedbackVote != "" {
		preds = append(preds, "m.feedback_vote = ?")
		args = append(args, f.FeedbackVote)
	}
	if f.FeedbackStatus != "" {
		preds = append(preds, "m.feedback_status = ?")
		args = append(args, f.FeedbackStatus)
	}
	from := "memory m"
	if f.Q != "" {
		if db.hasMemoryFTS() {
			// FTS5 MATCH join on the standalone memory_fts index, keyed back
			// to memory by rel_path.
			from = "memory m JOIN memory_fts ON memory_fts.rel_path = m.rel_path"
			preds = append(preds, "memory_fts MATCH ?")
			args = append(args, memoryFTSQuery(f.Q))
		} else {
			preds = append(preds, "m.body LIKE '%' || ? || '%'")
			args = append(args, f.Q)
		}
	}
	where := ""
	if len(preds) > 0 {
		where = " WHERE " + strings.Join(preds, " AND ")
	}
	cols := strings.ReplaceAll(memoryCols, "\n\t", " ")
	q := "SELECT " + prefixCols(cols, "m.") + " FROM " + from + where +
		" ORDER BY m.date DESC, m.rel_path"
	rows, err := db.getReader().QueryContext(ctx, q, args...)
	if err != nil {
		return nil, fmt.Errorf("listing memories: %w", err)
	}
	defer rows.Close()
	out := make([]Memory, 0, 64)
	for rows.Next() {
		m, err := scanMemory(rows)
		if err != nil {
			return nil, fmt.Errorf("scan memory: %w", err)
		}
		out = append(out, m)
	}
	return out, rows.Err()
}

// GetMemory returns a single memory note by its rel_path, or nil if not
// found.
func (db *DB) GetMemory(
	ctx context.Context, relPath string,
) (*Memory, error) {
	q := "SELECT " + memoryCols + " FROM memory WHERE rel_path = ?"
	rows, err := db.getReader().QueryContext(ctx, q, relPath)
	if err != nil {
		return nil, fmt.Errorf("getting memory: %w", err)
	}
	defer rows.Close()
	if !rows.Next() {
		return nil, rows.Err()
	}
	m, err := scanMemory(rows)
	if err != nil {
		return nil, fmt.Errorf("scan memory: %w", err)
	}
	return &m, nil
}

func (db *DB) MemoryEmbeddings(
	ctx context.Context, f MemoryFilter,
) ([]Memory, error) {
	var preds []string
	var args []any
	if f.Source != "" {
		preds = append(preds, "m.source = ?")
		args = append(args, f.Source)
	}
	if f.ProblemType != "" {
		preds = append(preds, "m.problem_type = ?")
		args = append(args, f.ProblemType)
	}
	if f.Type != "" {
		preds = append(preds, "m.type = ?")
		args = append(args, f.Type)
	}
	if f.Status != "" {
		preds = append(preds, "m.status = ?")
		args = append(args, f.Status)
	}
	if f.OriginSession != "" {
		preds = append(preds, "m.origin_session = ?")
		args = append(args, f.OriginSession)
	}
	if f.OriginProject != "" {
		preds = append(preds, "m.origin_project = ?")
		args = append(args, f.OriginProject)
	}
	if f.FeedbackVote != "" {
		preds = append(preds, "m.feedback_vote = ?")
		args = append(args, f.FeedbackVote)
	}
	if f.FeedbackStatus != "" {
		preds = append(preds, "m.feedback_status = ?")
		args = append(args, f.FeedbackStatus)
	}
	from := "memory m"
	if f.Q != "" {
		if db.hasMemoryFTS() {
			from = "memory m JOIN memory_fts ON memory_fts.rel_path = m.rel_path"
			preds = append(preds, "memory_fts MATCH ?")
			args = append(args, memoryFTSQuery(f.Q))
		} else {
			preds = append(preds, "m.body LIKE '%' || ? || '%'")
			args = append(args, f.Q)
		}
	}
	where := ""
	if len(preds) > 0 {
		where = " WHERE " + strings.Join(preds, " AND ")
	}
	cols := strings.ReplaceAll(memoryCols+", "+memoryEmbeddingCols, "\n\t", " ")
	q := "SELECT " + prefixCols(cols, "m.") + " FROM " + from + where +
		" ORDER BY m.date DESC, m.rel_path"
	rows, err := db.getReader().QueryContext(ctx, q, args...)
	if err != nil {
		return nil, fmt.Errorf("listing memory embeddings: %w", err)
	}
	defer rows.Close()
	return scanMemoryWithEmbedding(rows)
}

// prefixCols prefixes each bare column in a comma-separated list with the
// given table alias prefix. Used so the FTS join query can disambiguate
// the memory columns from the memory_fts columns.
func prefixCols(cols, prefix string) string {
	parts := strings.Split(cols, ",")
	for i, p := range parts {
		parts[i] = prefix + strings.TrimSpace(p)
	}
	return strings.Join(parts, ", ")
}

var memoryFTSTokenPattern = regexp.MustCompile(`[\p{L}\p{N}_]+`)

func memoryFTSQuery(q string) string {
	terms := memoryFTSTokenPattern.FindAllString(q, -1)
	if len(terms) == 0 {
		return quoteFTSTerm(q)
	}
	quoted := make([]string, 0, len(terms))
	for _, term := range terms {
		quoted = append(quoted, quoteFTSTerm(term))
	}
	return strings.Join(quoted, " AND ")
}

func quoteFTSTerm(term string) string {
	return `"` + strings.ReplaceAll(term, `"`, `""`) + `"`
}

// replaceMemoriesTx full-replaces the memory table inside an open tx. The
// AFTER INSERT/DELETE triggers keep memory_fts in sync automatically. When
// source is non-empty only that data source's rows are cleared and replaced,
// leaving the other source's rows untouched (the single memory table holds
// both cross-agent and cc-native rows, each synced by its own syncer).
func replaceMemoriesTx(
	ctx context.Context, tx txExec, source string, memories []Memory,
) error {
	if err := dropMemoryFTSMirrorIfBroken(ctx, tx); err != nil {
		return err
	}
	if source != "" {
		if _, err := tx.ExecContext(ctx,
			"DELETE FROM memory WHERE source = ?", source,
		); err != nil {
			return fmt.Errorf("clearing memory source %q: %w", source, err)
		}
	} else if _, err := tx.ExecContext(ctx, "DELETE FROM memory"); err != nil {
		return fmt.Errorf("clearing memory: %w", err)
	}
	cols := strings.ReplaceAll(memoryCols+", "+memoryEmbeddingCols, "\n\t", " ")
	stmt, err := tx.PrepareContext(ctx, `INSERT INTO memory (`+cols+`)
		VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)`)
	if err != nil {
		return fmt.Errorf("prepare memory insert: %w", err)
	}
	defer stmt.Close()
	for _, m := range memories {
		var encoded []byte
		var dim int
		if len(m.LLMEmbedding) > 0 {
			var err error
			encoded, err = EncodeEmbedding(m.LLMEmbedding)
			if err != nil {
				return fmt.Errorf("encode memory embedding %q: %w", m.RelPath, err)
			}
			dim = len(m.LLMEmbedding)
		}
		if _, err := stmt.ExecContext(ctx,
			m.RelPath, m.Source, m.Title, m.Date, m.ProblemType, m.Type,
			m.Status, m.OriginSession, m.OriginProject, m.FeedbackVote,
			m.FeedbackComment, m.FeedbackStatus, m.CanonicalCoveredRefs,
			m.CanonicalProvenance, m.Body, m.BodyTokens, m.SourceMtime,
			m.SyncedAt, encoded, dim,
		); err != nil {
			return fmt.Errorf("insert memory %q: %w", m.RelPath, err)
		}
	}
	return nil
}

func dropMemoryFTSMirrorIfBroken(ctx context.Context, tx txExec) error {
	if _, err := tx.ExecContext(ctx, "SELECT 1 FROM memory_fts LIMIT 1"); err == nil {
		return nil
	}
	stmts := []string{
		"DROP TRIGGER IF EXISTS memory_ai",
		"DROP TRIGGER IF EXISTS memory_ad",
		"DROP TRIGGER IF EXISTS memory_au",
		"DROP TABLE IF EXISTS memory_fts",
	}
	for _, stmt := range stmts {
		if _, err := tx.ExecContext(ctx, stmt); err != nil {
			if strings.HasPrefix(stmt, "DROP TABLE") && strings.Contains(err.Error(), "no such module") {
				continue
			}
			return fmt.Errorf("dropping broken memory fts mirror (%s): %w", stmt, err)
		}
	}
	return nil
}

// ReplaceMemories atomically full-replaces the whole memory table in a single
// transaction (both data sources). It is retained for the combined-replace
// path; the per-source syncers use ReplaceMemoriesBySource so each source can
// sync independently without wiping the other. Local-only writer (PG/DuckDB
// receive rows via the SQLite mirror, not through this path).
func (db *DB) ReplaceMemories(
	ctx context.Context, memories []Memory,
) error {
	return db.replaceMemories(ctx, "", memories)
}

// ReplaceMemoriesBySource atomically replaces only the rows of the given data
// source (cross-agent or cc-native), leaving the other source untouched. Each
// memory syncer owns one source and calls this so the two syncers — sharing
// the single memory table — never clobber each other.
func (db *DB) ReplaceMemoriesBySource(
	ctx context.Context, source string, memories []Memory,
) error {
	return db.replaceMemories(ctx, source, memories)
}

func (db *DB) replaceMemories(
	ctx context.Context, source string, memories []Memory,
) error {
	db.mu.Lock()
	defer db.mu.Unlock()
	tx, err := db.getWriter().BeginTx(ctx, nil)
	if err != nil {
		return fmt.Errorf("begin memory tx: %w", err)
	}
	defer func() { _ = tx.Rollback() }()
	if err := replaceMemoriesTx(ctx, tx, source, memories); err != nil {
		return err
	}
	if err := tx.Commit(); err != nil {
		return fmt.Errorf("commit memory: %w", err)
	}
	return nil
}

File: frontend/src/lib/api/memory.ts
import { getBase, authHeaders, ApiError, responseErrorMessage } from "./runtime";

// Memory mirrors db.Memory: one user-memory note (a markdown file under the
// memory SSOT) with YAML frontmatter fields plus a body and sync metadata.
export interface Memory {
  rel_path: string;
  // Data source: "cross-agent" (the user-memory SSOT) or "cc-native"
  // (CC auto-memory scanned across project dirs).
  source: string;
  title: string;
  date: string;
  problem_type: string;
  type: string;
  status: string;
  origin_session: string;
  // Project a note belongs to ("" = the General bucket: user-global or
  // cross-project notes). Drives the project facet/grouping.
  origin_project: string;
  feedback_vote: string;
  feedback_comment: string;
  feedback_status: string;
  body?: string;
  body_tokens: number;
  source_mtime: number;
  synced_at: string;
}

// MemoryFilter narrows the listing. Empty / undefined fields = no filter. q is
// a full-text query over the note body.
export interface MemoryFilter {
  source?: string;
  problem_type?: string;
  type?: string;
  status?: string;
  origin_session?: string;
  origin_project?: string;
  feedback_vote?: string;
  feedback_comment?: string;
  feedback_status?: string;
  q?: string;
}

async function getJSON<T>(path: string, signal?: AbortSignal): Promise<T> {
  const res = await fetch(`${getBase()}${path}`, authHeaders({ signal }));
  if (!res.ok) {
    throw new ApiError(res.status, await responseErrorMessage(res));
  }
  return (await res.json()) as T;
}

async function sendJSON<T>(
  method: "PUT" | "POST",
  path: string,
  body: unknown,
  signal?: AbortSignal,
): Promise<T> {
  const res = await fetch(
    `${getBase()}${path}`,
    authHeaders({
      method,
      signal,
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
  );
  if (!res.ok) {
    throw new ApiError(res.status, await responseErrorMessage(res));
  }
  return (await res.json()) as T;
}

export async function fetchMemories(
  filter: MemoryFilter = {},
  signal?: AbortSignal,
): Promise<Memory[]> {
  const params = new URLSearchParams();
  if (filter.source) params.set("source", filter.source);
  if (filter.problem_type) params.set("problem_type", filter.problem_type);
  if (filter.type) params.set("type", filter.type);
  if (filter.status) params.set("status", filter.status);
  if (filter.origin_session) params.set("origin_session", filter.origin_session);
  if (filter.origin_project) params.set("origin_project", filter.origin_project);
  if (filter.feedback_vote) params.set("feedback_vote", filter.feedback_vote);
  if (filter.feedback_status) params.set("feedback_status", filter.feedback_status);
  if (filter.q) params.set("q", filter.q);
  const qs = params.toString();
  const path = qs ? `/memories?${qs}` : "/memories";
  const body = await getJSON<{ memories: Memory[] }>(path, signal);
  return body.memories ?? [];
}

// fetchMemory loads one note by its rel_path. The backend keys notes by
// rel_path which contains slashes, so the path is URL-safe base64 encoded
// (matching the server's base64.RawURLEncoding decode).
export function fetchMemory(
  relPath: string,
  signal?: AbortSignal,
): Promise<Memory> {
  return getJSON<Memory>(`/memories/${encodeMemoryPath(relPath)}`, signal);
}

// MemoryRaw is the verbatim on-disk file content plus its sha256. The editor
// loads this (not the DB-parsed Memory) so it can round-trip untracked
// frontmatter keys and obtain a base_sha that matches the write gate.
export interface MemoryRaw {
  content: string;
  sha: string;
}

// fetchMemoryRaw loads a note's raw file content and sha for editing.
export function fetchMemoryRaw(
  relPath: string,
  signal?: AbortSignal,
): Promise<MemoryRaw> {
  return getJSON<MemoryRaw>(
    `/memories/${encodeMemoryPath(relPath)}/raw`,
    signal,
  );
}

// putMemory writes back the full reconstructed file content. base_sha is the
// sha the editor read at edit time; a stale base yields a 409 (ApiError with
// status 409) which the caller must surface, never silently drop.
export function putMemory(
  relPath: string,
  content: string,
  baseSHA: string,
  signal?: AbortSignal,
): Promise<{ sha: string }> {
  return sendJSON<{ sha: string }>(
    "PUT",
    `/memories/${encodeMemoryPath(relPath)}`,
    { content, base_sha: baseSHA },
    signal,
  );
}

export interface MemoryFeedbackInput {
  vote: "up" | "down" | "";
  comment: string;
  status: "pending" | "handled" | "";
  base_sha?: string;
}

export function setMemoryFeedback(
  relPath: string,
  feedback: MemoryFeedbackInput,
  signal?: AbortSignal,
): Promise<{ sha: string }> {
  return sendJSON<{ sha: string }>(
    "POST",
    `/memories/${encodeMemoryPath(relPath)}/feedback`,
    feedback,
    signal,
  );
}

// MemoryHistoryEntry mirrors memory.HistoryEntry: one git commit touching the
// note, newest first.
export interface MemoryHistoryEntry {
  commit: string;
  date: string;
  message: string;
}

// fetchMemoryHistory lists the git history for a note (empty when the memory
// dir is not a git repo or the file was never committed).
export async function fetchMemoryHistory(
  relPath: string,
  signal?: AbortSignal,
): Promise<MemoryHistoryEntry[]> {
  const body = await getJSON<{ history: MemoryHistoryEntry[] }>(
    `/memories/${encodeMemoryPath(relPath)}/history`,
    signal,
  );
  return body.history ?? [];
}

// fetchMemoryAtCommit returns the note's content at a specific commit, for the
// history diff view.
export async function fetchMemoryAtCommit(
  relPath: string,
  commit: string,
  signal?: AbortSignal,
): Promise<string> {
  const body = await getJSON<{ content: string }>(
    `/memories/${encodeMemoryPath(relPath)}/history/${encodeURIComponent(commit)}`,
    signal,
  );
  return body.content ?? "";
}

// revertMemory restores the note to a past commit. base_sha guards against a
// concurrent on-disk change between viewing and reverting (409 on mismatch).
export function revertMemory(
  relPath: string,
  commit: string,
  baseSHA: string,
  signal?: AbortSignal,
): Promise<{ sha: string }> {
  return sendJSON<{ sha: string }>(
    "POST",
    `/memories/${encodeMemoryPath(relPath)}/revert`,
    { commit, base_sha: baseSHA },
    signal,
  );
}

// encodeMemoryPath produces URL-safe base64 without padding, matching Go's
// base64.RawURLEncoding so the server can round-trip the rel_path.
function encodeMemoryPath(relPath: string): string {
  const bytes = new TextEncoder().encode(relPath);
  let bin = "";
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}
